src/algebraic_geometry/Gamma_Spec_adjunction.lean

```diff
@@ -0,0 +1,273 @@
+  have hr : is_unit (X.Γ_to_stalk x r),
+    apply_fun sm at he,
+    rw [←to_stalk_comm, comp_apply],
+    erw ← he, rw ring_hom.map_mul,
+    apply is_unit.mul ht,
+    exact is_unit.map sm.to_monoid_hom (hu s),
+  rw ← not_mem_prime_iff_unit_in_stalk at hr,
+  have hr' := hu ⟨r,hr⟩, erw ← he at hr',
+  exact is_unit_of_mul_is_unit_left hr',
+end
+
+lemma comp_ring_hom_ext {X : LocallyRingedSpace} {R : CommRing}
+  {f : R ⟶ Γ' X} {β : X ⟶ Spec.LocallyRingedSpace_obj R}
+  (w : X.to_Γ_Spec.1.1 ≫ (Spec.LocallyRingedSpace_map f).1.1 = β.1.1)
+  (h : ∀ r : R, f ≫ X.presheaf.map (@le_top _ _ ((opens.map β.1.1).obj (basic_open r))).hom.op
+                        /- slower : by simp; exact le_top.hom.op -/
+              = to_open R (basic_open r) ≫ β.1.c.app (op (basic_open r))) :
+  X.to_Γ_Spec ≫ Spec.LocallyRingedSpace_map f = β :=
+begin
+  ext1, apply Spec.hom_ext,
+  { intros r _,
+    rw LocallyRingedSpace.comp_val_c_app',
+    erw to_open_comp_comap_assoc,
+    rw category.assoc,
+    erw [to_Γ_Spec_SheafedSpace_app_spec, ←X.presheaf.map_comp],
+    convert h r },
+  exact w,
+end
+
+/-- `to_Spec_Γ _` is iso so these are mutually two-sided inverses. -/
+lemma left_triangle : to_Spec_Γ (Γ' X) ≫ X.to_Γ_Spec.1.c.app (op ⊤) = 𝟙 (Γ' X) :=
+begin
+  unfold to_Spec_Γ,
+  rw ← to_open_res _ (basic_open (1 : Γ' X)) ⊤ (eq_to_hom basic_open_one.symm),
+  erw category.assoc, rw [nat_trans.naturality, ←category.assoc],
+  erw [X.to_Γ_Spec_SheafedSpace_app_spec 1, ←functor.map_comp],
+  convert eq_to_hom_map X.presheaf _, refl,
+end
+
+end LocallyRingedSpace
+
+/-- Unit as a natural transformation. -/
+def identity_to_Γ_Spec : 𝟭 LocallyRingedSpace.{u} ⟶ Γ.right_op ⋙ Spec.to_LocallyRingedSpace :=
+{ app := LocallyRingedSpace.to_Γ_Spec,
+  naturality' := λ X Y f, begin
+    apply eq.symm, apply LocallyRingedSpace.comp_ring_hom_ext,
+    { ext1 x, apply eq.symm, convert congr_fun (congr_arg comap
+        (PresheafedSpace.stalk_map_germ f.1 ⊤ ⟨x,trivial⟩))
+        (@local_ring.closed_point _ _ (X.local_ring x)),
+      erw prime_spectrum.comap_comp, rw function.comp_apply,
+      erw (@local_ring.local_hom_iff_comap_closed_point
+        _ _ (Y.2 _) _ _ (X.2 x) _).1 (f.2 x), refl },
+    { intro r, rw [LocallyRingedSpace.comp_val_c_app', ←category.assoc],
+      erw [Y.to_Γ_Spec_SheafedSpace_app_spec, f.1.c.naturality], refl },
+  end }
+
+namespace Γ_Spec
+
+/-- `Spec_Γ_identity` is iso so these are mutually two-sided inverses. -/
+lemma right_triangle (R : CommRing) :
+  identity_to_Γ_Spec.app (Spec.LocallyRingedSpace_obj R) ≫
+  Spec.LocallyRingedSpace_map (Spec_Γ_identity.inv.app R) = 𝟙 _ :=
+begin
+  apply LocallyRingedSpace.comp_ring_hom_ext,
+  { ext1 p, ext, erw ← @is_localization.at_prime.to_map_mem_maximal_iff _ _ _ _
+    (to_stalk R p).to_algebra p.1 _ (is_localization.to_stalk R p) x, refl },
+  { intro r, apply to_open_res },
+end
+
+/-- Auxiliary data structure for defining the adjunction. -/
+def core_unit_counit :
```

Comment:
Could/Should this be inlined in the definition of `adjunction`? 
If not (timeout issues?), I think it may need `simps` or some simp lemmas so that the api of `adjunction` can be used easily without manually unfolding definitions.

Reply:
OK, it's now been inlined.